chalice/deploy/deployer.py

```diff
@@ -169,6 +177,104 @@ def _validate_manage_iam_role(config):
             )
 
 
+class ChaliceDeploymentError(Exception):
+    def __init__(self, error):
+        # type: (Any) -> None
+        where = self._get_error_location(error)
+        msg = self._wrap_text(
+            'ERROR - %s, received the following error:' % where
+        )
+        msg += '\n\n'
+        msg += self._wrap_text(self._get_error_message(error), indent=' ')
+        msg += '\n\n'
+        suggestion = self._get_error_suggestion(error)
+        if suggestion:
+            msg += self._wrap_text(suggestion)
+        super(ChaliceDeploymentError, self).__init__(msg)
+
+    def _get_error_location(self, error):
+        # type: (Any) -> str
+        where = 'While deploying your chalice application'
+        if isinstance(error, LambdaClientError):
+            where = (
+                'While sending your chalice handler code to Lambda function '
+                '"%s"' % error.context.function_name
+            )
+        return where
+
+    def _get_error_message(self, error):
+        # type: (Any) -> str
+        msg = str(error)
+        if isinstance(error, LambdaClientError):
+            if isinstance(error.original_error, RequestsConnectionError):
+                msg = self._get_error_message_for_connection_error(
+                    error.original_error)
+        return msg
+
+    def _get_error_message_for_connection_error(self, connection_error):
+        # type: (RequestsConnectionError) -> str
+
+        # To get the underlying error that raised the
+        # requests.ConnectionError it is required to go down two levels of
+        # arguments to get the underlying exception. The instantiation of
+        # one of these exceptions looks like this:
+        #
+        # requests.ConnectionError(
+        #     urllib3.exceptions.ProtocolError(
+        #         'Connection aborted.', <SomeException>)
+        # )
+        message = connection_error.args[0].args[0]
+        underlying_error = connection_error.args[0].args[1]
+
+        # In python3, this is a BrokenPipeError. However in python2, this
+        # is a socket.error that has the message 'Broken pipe' in it. So we
+        # don't want to be assuming all socket.error are broken pipes so just
+        # check if the message has 'Broken pipe' in it.
+        if 'Broken pipe' in str(underlying_error):
```

Comment:
This seems more appropriately handled in `compat.py`.  Here we actually have a solid `isinstance` way of checking for broken pipe in python3, and it seems to make sense to use it if possible rather than rely on `__contains__`.